File: problems/dynamic_programming/maximum_sum_submatrix/js/maximum_sum_submatrix.js
export const maximumSumSubmatrix = (() => {
  const buildDp = (matrix) => {
    const rows = matrix.length;
    const cols = matrix[0].length;

    const dp = Array.from({ length: rows + 1 }, () => Array(cols + 1).fill(0));

    for (let i = 1; i <= rows; i++) {
      for (let j = 1; j <= cols; j++) {
        dp[i][j] =
          matrix[i - 1][j - 1] + dp[i - 1][j] + dp[i][j - 1] - dp[i - 1][j - 1];
      }
    }
    return dp;
  };

  const findMaxSum = (dp, size, dimensions) => {
    let maxSum = -Infinity;
    const [rows, cols] = dimensions;

    for (let i = 1; i <= rows - size + 1; i++) {
      for (let j = 1; j <= cols - size + 1; j++) {
        const sum =
          dp[i + size - 1][j + size - 1] -
          dp[i - 1][j + size - 1] -
          dp[i + size - 1][j - 1] +
          dp[i - 1][j - 1];

        if (sum > maxSum) maxSum = sum;
      }
    }
    return maxSum;
  };

  return (matrix, size) => {
    const dp = buildDp(matrix);

    return findMaxSum(dp, size, [matrix.length, matrix[0].length]);
  };
})();

File: problems/dynamic_programming/maximum_sum_submatrix/cpp/maximum_sum_submatrix.cpp
#include "maximum_sum_submatrix.h"
#include <climits>

vector2DInt buildDp(const vector2DInt &matrix) {
  int rows = matrix.size();
  int cols = matrix[0].size();

  vector2DInt dp(rows + 1, std::vector<int>(cols + 1, 0));

  for (int i = 1; i <= rows; i++) {
    for (int j = 1; j <= cols; j++) {
      dp[i][j] =
          matrix[i - 1][j - 1] + dp[i - 1][j] + dp[i][j - 1] - dp[i - 1][j - 1];
    }
  }
  return dp;
}

int findMaxSum(const vector2DInt &dp, int size, const Dimensions &dimensions) {
  int max_sum = INT_MIN;

  for (int i = 1; i <= dimensions.first - size + 1; i++) {
    for (int j = 1; j <= dimensions.second - size + 1; j++) {
      int sum = dp[i + size - 1][j + size - 1] - dp[i - 1][j + size - 1] -
                dp[i + size - 1][j - 1] + dp[i - 1][j - 1];

      if (sum > max_sum)
        max_sum = sum;
    }
  }
  return max_sum;
}
int maximumSumSubmatrix(vector2DInt matrix, int size) {
  vector2DInt dp = buildDp(matrix);

  return findMaxSum(dp, size, {matrix.size(), matrix[0].size()});
}
